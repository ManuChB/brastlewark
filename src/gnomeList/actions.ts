import { SAVE_DATA, SHOW_DETAILS, CLOSE_MODAL, SET_FILTERED_DATA} from './types';
import axios from 'axios';
import { Dispatch, Action } from 'redux';
import { IGnomeProp } from '../gnome/gnome';
import { IGnomeListActions } from './gnomeList';
import { IProfessionsProp } from '../professions/professions';
import _ from 'lodash';

const actions: IGnomeListActions = {
    saveData(gnomes: Array<IGnomeProp>) {
        return {
            type: SAVE_DATA,
            payload: gnomes
        }
    },
    showGnomeDetails(gnome: IGnomeProp, friends: Array<IGnomeProp>) {
        return {
            type: SHOW_DETAILS,
            payload: {gnome, friends}
        }
    },
    closeModal() {
        return {
            type: CLOSE_MODAL
        }
    },
    setFilteredData(gnomes: Array<IGnomeProp>) {
        return {
            type: SET_FILTERED_DATA,
            payload: gnomes
        }
    }
}


export const fetchData = () => {

    return ( dispatch: Dispatch<Action<any>>) =>  {
        axios.get(
            `https://raw.githubusercontent.com/rrafols/mobile_test/master/data.json`
        )
            .then(response => {
                dispatch(actions.saveData(response.data.Brastlewark));
                return response;
            })
            .catch(error => {
                return `[gnomeList][action][fetchData][error]>>> ${error}`;
            });
    }
}

const getProfessions = (gnomes: Array<IGnomeProp>) => {
    let profesionList: Array<string> = [];
    gnomes.forEach( gnome => {
        if(!_.isEmpty(gnome.professions)) {
            gnome.professions.forEach(profession => {
                if (!profesionList.includes(profession)) {
                    profesionList.push(profession);
                }
            })
        }
    })
}

export default actions;
