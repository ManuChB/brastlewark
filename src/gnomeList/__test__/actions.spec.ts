import actions from '../actions';

describe('Actions', () => {
    const gnome = {
        id: 1,
        name: 'name',
        thumbnail: 'image',
        age: 100,
        weight: 100,
        height: 100,
        hair_color: 'red',
        professions: [],
        friends: [],
        onPress: jest.fn()
    }
    const gnomeList = [gnome]

    it('saveData() should return action', () => {
        expect(actions.saveData(gnomeList)).toMatchSnapshot();
    });
    it('showGnomeDetails() should return action', () => {
        expect(actions.showGnomeDetails(gnome, gnomeList)).toMatchSnapshot();
    });
    it('closeModal() should return action', () => {
        expect(actions.closeModal()).toMatchSnapshot();
    });
    it('setFilteredData() should return action', () => {
        expect(actions.setFilteredData(gnomeList)).toMatchSnapshot();
    });
});
