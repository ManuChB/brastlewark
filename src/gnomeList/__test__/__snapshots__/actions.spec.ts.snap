// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Actions closeModal() should return action 1`] = `
Object {
  "type": "@@main/CLOSE_MODAL",
}
`;

exports[`Actions saveData() should return action 1`] = `
Object {
  "payload": Array [
    Object {
      "age": 100,
      "friends": Array [],
      "hair_color": "red",
      "height": 100,
      "id": 1,
      "name": "name",
      "onPress": [MockFunction],
      "professions": Array [],
      "thumbnail": "image",
      "weight": 100,
    },
  ],
  "type": "@@main/SAVE_DATA",
}
`;

exports[`Actions setFilteredData() should return action 1`] = `
Object {
  "payload": Array [
    Object {
      "age": 100,
      "friends": Array [],
      "hair_color": "red",
      "height": 100,
      "id": 1,
      "name": "name",
      "onPress": [MockFunction],
      "professions": Array [],
      "thumbnail": "image",
      "weight": 100,
    },
  ],
  "type": "@@main/SET_FILTERED_DATA",
}
`;

exports[`Actions showGnomeDetails() should return action 1`] = `
Object {
  "payload": Object {
    "friends": Array [
      Object {
        "age": 100,
        "friends": Array [],
        "hair_color": "red",
        "height": 100,
        "id": 1,
        "name": "name",
        "onPress": [MockFunction],
        "professions": Array [],
        "thumbnail": "image",
        "weight": 100,
      },
    ],
    "gnome": Object {
      "age": 100,
      "friends": Array [],
      "hair_color": "red",
      "height": 100,
      "id": 1,
      "name": "name",
      "onPress": [MockFunction],
      "professions": Array [],
      "thumbnail": "image",
      "weight": 100,
    },
  },
  "type": "@@main/SHOW_DETAILS",
}
`;
